htlcswitch/link_test.go

```diff
@@ -4500,6 +4634,170 @@ func TestChannelLinkWaitForRevocation(t *testing.T) {
 	}
 }
 
+// TestChannelLinkBatchPreimageWrite asserts that a link will batch preimage
+// writes when just as it receives a CommitSig to lock in any Settles, and also
+// if the link is aware of any uncommitted preimages if the link is stopped,
+// i.e. due to a disconnection or shutdown.
+func TestChannelLinkBatchPreimageWrite(t *testing.T) {
+	t.Parallel()
+
+	const chanAmt = btcutil.SatoshiPerBitcoin * 5
+	const chanReserve = btcutil.SatoshiPerBitcoin * 1
+	aliceLink, bobChannel, batchTicker, startUp, cleanUp, restore, err :=
+		newSingleLinkTestHarness(chanAmt, chanReserve)
+	if err != nil {
+		t.Fatalf("unable to create link: %v", err)
+	}
+	defer cleanUp()
+
+	if err := startUp(); err != nil {
+		t.Fatalf("unable to start test harness: %v", err)
+	}
+
+	alice := newPersistentLinkHarness(t, aliceLink, batchTicker, restore)
+
+	// We will send 10 HTLCs in total, from Bob to Alice.
+	numHtlcs := 10
+	var htlcs []*lnwire.UpdateAddHTLC
+	var invoices []*channeldb.Invoice
+	for i := 0; i < numHtlcs; i++ {
+		htlc, invoice := generateHtlcAndInvoice(t, uint64(i))
+		htlcs = append(htlcs, htlc)
+		invoices = append(invoices, invoice)
+	}
+
+	// The test will proceed in two iterations, where we send half of the
+	// htlcs on the first pass and the latter half on the second.
+	batchSize := numHtlcs / 2
+
+	// Initialize the parameters for the first run of the tests. The start
+	// and end parameters are indices pointing into the htlcs and invoices
+	// array, and will be shifted by batchSize after the first pass. The
+	// disconnect bool specifies whether the test should wait for the
+	// CommitSig covering the settles from Bob. On the first pass, we will
+	// go for a full revocation dance and ensure the batch of preimages was
+	// committed. On the second, we'll disconnect before allowing Alice to
+	// process the CommitSig, and ensure that the link writes the preimage
+	// batch before exiting.
+	var (
+		start      = 0
+		end        = batchSize
+		disconnect = false
+	)
+
+checkBatchConsistency:
```

Comment:
I think it would be more understandable to use one or two functions instead of the `goto`. Or alternatively make it a parameterized test (with and without disconnect) that is ran from a parent test using `t.Run`.

Reply:
good suggestion, i'm actually realizing now that the persistent link harness is unnecessary for this test, so splitting it into a table driven test would be easier now